File: swiftui/CellposeGUI/Services/SeriesDiscoveryService.swift
import Foundation

final class SeriesDiscoveryService {
    private let supportedExtensions = ["jpg", "jpeg", "png", "tif", "tiff"]
    private let supportedSubfolderTemplates = ["Pos{p}", "Position{p}", "Pos_{p}", "Position_{p}"]
    private let supportedFilenameTemplates = ["img_{t}_{c}_{z}.jpg", "img_channel{c}_position{p}_time{t}_z{z}.tif"]

    func suggestTemplates(folder: String) throws -> SeriesTemplateSuggestion {
        guard FileManager.default.fileExists(atPath: folder) else {
            throw SidecarError.serverError("Folder not found: \(folder)")
        }

        var best: (String, String, Int)?
        for subfolder in supportedSubfolderTemplates {
            for filename in supportedFilenameTemplates {
                let matches = collectMatches(folder: folder, subfolderTemplate: subfolder, filenameTemplate: filename)
                guard !matches.isEmpty else { continue }
                if best == nil || matches.count > best!.2 {
                    best = (subfolder, filename, matches.count)
                }
            }
        }

        if let best {
            return SeriesTemplateSuggestion(subfolderTemplate: best.0, filenameTemplate: best.1)
        }
        return SeriesTemplateSuggestion(subfolderTemplate: "", filenameTemplate: "img_{t}_{c}_{z}.jpg")
    }

    func discover(folder: String, subfolderTemplate: String, filenameTemplate: String) throws -> SeriesDiscovery {
        var matches = collectMatches(folder: folder, subfolderTemplate: subfolderTemplate, filenameTemplate: filenameTemplate)
        guard !matches.isEmpty else {
            throw SidecarError.serverError("No files matched template in \(folder)")
        }

        matches.sort { lhs, rhs in
            let left = (
                sortRank(lhs.position),
                sortRank(lhs.time),
                sortRank(lhs.channel),
                sortRank(lhs.z),
                lhs.relativePath
            )
            let right = (
                sortRank(rhs.position),
                sortRank(rhs.time),
                sortRank(rhs.channel),
                sortRank(rhs.z),
                rhs.relativePath
            )
            return left < right
        }

        var lookup: [String: Int] = [:]
        var records: [SeriesDiscoveryRecord] = []
        var datasetRecords: [SeriesDatasetRecord] = []
        for (index, match) in matches.enumerated() {
            let key = "\(match.position)_\(match.time)_\(match.channel)_\(match.z)"
            guard lookup[key] == nil else {
                throw SidecarError.serverError("Duplicate series record for \(key)")
            }
            lookup[key] = index
            records.append(SeriesDiscoveryRecord(index: index, label: match.relativePath, path: match.path))
            datasetRecords.append(
                SeriesDatasetRecord(
                    label: match.relativePath,
                    path: match.path,
                    position: match.position,
                    time: match.time,
                    channel: match.channel,
                    z: match.z
                )
            )
        }

        let axes = [
            "position": uniqueSorted(matches.map(\.position)),
            "time": uniqueSorted(matches.map(\.time)),
            "channel": uniqueSorted(matches.map(\.channel)),
            "z": uniqueSorted(matches.map(\.z)),
        ]
        var axisIndex: [String: [String: Int]] = [:]
        for (axis, values) in axes {
            axisIndex[axis] = Dictionary(uniqueKeysWithValues: values.enumerated().map { ($1, $0) })
        }

        return SeriesDiscovery(
            folder: folder,
            recordCount: records.count,
            axes: axes,
            records: records,
            dataset: SeriesDatasetPayload(
                folder: folder,
                subfolderTemplate: subfolderTemplate,
                filenameTemplate: filenameTemplate,
                axes: axes,
                axisIndex: axisIndex,
                lookup: lookup,
                records: datasetRecords
            )
        )
    }

    private func uniqueSorted(_ values: [String]) -> [String] {
        Array(Set(values)).sorted { sortRank($0) < sortRank($1) }
    }

    private func sortRank(_ value: String) -> (Int, String) {
        if let number = Int(value) {
            return (0, String(format: "%010d", number))
        }
        return (1, value)
    }

    private func collectMatches(folder: String, subfolderTemplate: String, filenameTemplate: String) -> [SeriesMatch] {
        let subfolderPattern = compileTemplate(subfolderTemplate, allowEmpty: true)
        let filenamePattern = compileTemplate(filenameTemplate)
        var matches: [SeriesMatch] = []
        let folderURL = URL(fileURLWithPath: folder)
        let candidateFolders: [URL]
        if subfolderTemplate.isEmpty {
            candidateFolders = [folderURL]
        } else {
            candidateFolders = (try? FileManager.default.contentsOfDirectory(
                at: folderURL,
                includingPropertiesForKeys: nil
            ).filter(\.hasDirectoryPath).sorted(by: { $0.lastPathComponent < $1.lastPathComponent })) ?? []
        }

        for candidate in candidateFolders {
            let subfolderName = subfolderTemplate.isEmpty ? "" : candidate.lastPathComponent
            if !subfolderTemplate.isEmpty {
                guard subfolderPattern.firstMatch(in: subfolderName, range: NSRange(subfolderName.startIndex..., in: subfolderName)) != nil else {
                    continue
                }
            }

            let subfolderValues = namedGroups(from: subfolderName, pattern: subfolderPattern)
            let files = (try? FileManager.default.contentsOfDirectory(at: candidate, includingPropertiesForKeys: nil)
                .filter { !$0.hasDirectoryPath && isSupportedSeriesFile($0.lastPathComponent) }
                .sorted(by: { $0.lastPathComponent < $1.lastPathComponent })) ?? []

            for file in files {
                let name = file.lastPathComponent
                guard filenamePattern.firstMatch(in: name, range: NSRange(name.startIndex..., in: name)) != nil else {
                    continue
                }
                var values = subfolderValues
                for (key, value) in namedGroups(from: name, pattern: filenamePattern) {
                    values[key] = value
                }
                let relativePath = subfolderName.isEmpty ? name : "\(subfolderName)/\(name)"
                matches.append(
                    SeriesMatch(
                        path: file.path,
                        relativePath: relativePath,
                        position: values["position", default: "0"],
                        time: values["time", default: "0"],
                        channel: values["channel", default: "0"],
                        z: values["z", default: "0"]
                    )
                )
            }
        }
        return matches
    }

    private func compileTemplate(_ template: String, allowEmpty: Bool = false) -> NSRegularExpression {
        if template.isEmpty {
            guard allowEmpty else {
                fatalError("Filename template cannot be empty.")
            }
            return try! NSRegularExpression(pattern: "^$", options: [.caseInsensitive])
        }
        return try! NSRegularExpression(pattern: compileTemplatePattern(template), options: [.caseInsensitive])
    }

    private func compileTemplatePattern(_ template: String) -> String {
        var parts = ["^"]
        let nsTemplate = template as NSString
        let placeholderRegex = try! NSRegularExpression(pattern: #"\{([a-zA-Z_][a-zA-Z0-9_]*)\}"#)
        var searchStart = 0
        while searchStart < nsTemplate.length {
            let range = NSRange(location: searchStart, length: nsTemplate.length - searchStart)
            guard let match = placeholderRegex.firstMatch(in: template, options: [], range: range) else {
                parts.append(NSRegularExpression.escapedPattern(for: nsTemplate.substring(from: searchStart)))
                break
            }
            if match.range.location > searchStart {
                parts.append(NSRegularExpression.escapedPattern(for: nsTemplate.substring(with: NSRange(location: searchStart, length: match.range.location - searchStart))))
            }
            let name = nsTemplate.substring(with: match.range(at: 1))
            switch name {
            case "t", "time": parts.append("(?<time>.+?)")
            case "p", "position": parts.append("(?<position>.+?)")
            case "c", "channel": parts.append("(?<channel>.+?)")
            case "z": parts.append("(?<z>.+?)")
            default: fatalError("Unsupported placeholder {\(name)}")
            }
            searchStart = match.range.location + match.range.length
        }
        parts.append("$")
        return parts.joined()
    }

    private func namedGroups(from text: String, pattern: NSRegularExpression) -> [String: String] {
        guard let match = pattern.firstMatch(in: text, range: NSRange(text.startIndex..., in: text)) else { return [:] }
        var result: [String: String] = [:]
        for name in ["position", "time", "channel", "z"] {
            let range = match.range(withName: name)
            guard range.location != NSNotFound, let swiftRange = Range(range, in: text) else { continue }
            result[name] = String(text[swiftRange])
        }
        return result
    }

    private func isSupportedSeriesFile(_ name: String) -> Bool {
        if name.lowercased().hasSuffix("_seg.cellpose") { return false }
        let ext = URL(fileURLWithPath: name).pathExtension.lowercased()
        return supportedExtensions.contains(ext)
    }

    private struct SeriesMatch {
        let path: String
        let relativePath: String
        let position: String
        let time: String
        let channel: String
        let z: String
    }
}

File: winui/CellposeGUI/Services/SeriesDiscoveryService.cs
using System.IO.Compression;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Text.RegularExpressions;
using CellposeGUI.Models;

namespace CellposeGUI.Services;

public sealed class SeriesDiscoveryService
{
    private static readonly string[] SupportedExtensions = [".jpg", ".jpeg", ".png", ".tif", ".tiff"];
    private static readonly string[] SupportedSubfolderTemplates =
        ["Pos{p}", "Position{p}", "Pos_{p}", "Position_{p}"];
    private static readonly string[] SupportedFilenameTemplates =
        ["img_{t}_{c}_{z}.jpg", "img_channel{c}_position{p}_time{t}_z{z}.tif"];

    public SeriesTemplateSuggestion SuggestTemplates(string folder)
    {
        if (!Directory.Exists(folder))
            throw new SidecarException($"Folder not found: {folder}");

        (string Subfolder, string Filename, int Count)? best = null;
        foreach (var subfolder in SupportedSubfolderTemplates)
        {
            foreach (var filename in SupportedFilenameTemplates)
            {
                var matches = CollectMatches(folder, subfolder, filename);
                if (matches.Count == 0)
                    continue;
                var score = matches.Count;
                if (best == null || score > best.Value.Count)
                    best = (subfolder, filename, score);
            }
        }

        if (best != null)
            return new SeriesTemplateSuggestion { SubfolderTemplate = best.Value.Subfolder, FilenameTemplate = best.Value.Filename };

        return new SeriesTemplateSuggestion
        {
            SubfolderTemplate = "",
            FilenameTemplate = "img_{t}_{c}_{z}.jpg",
        };
    }

    public SeriesDiscovery Discover(string folder, string subfolderTemplate, string filenameTemplate)
    {
        var matches = CollectMatches(folder, subfolderTemplate, filenameTemplate);
        if (matches.Count == 0)
            throw new SidecarException($"No files matched template in {folder}");

        matches = matches
            .OrderBy(m => SortKey(m.Position))
            .ThenBy(m => SortKey(m.Time))
            .ThenBy(m => SortKey(m.Channel))
            .ThenBy(m => SortKey(m.Z))
            .ThenBy(m => m.RelativePath)
            .ToList();

        var lookup = new Dictionary<string, int>();
        var records = new List<SeriesDiscoveryRecord>();
        var datasetRecords = new List<SeriesDatasetRecord>();
        for (var i = 0; i < matches.Count; i++)
        {
            var match = matches[i];
            var key = $"{match.Position}_{match.Time}_{match.Channel}_{match.Z}";
            if (lookup.ContainsKey(key))
                throw new SidecarException($"Duplicate series record for {key}");
            lookup[key] = i;
            records.Add(new SeriesDiscoveryRecord { Index = i, Label = match.RelativePath, Path = match.Path });
            datasetRecords.Add(new SeriesDatasetRecord
            {
                Label = match.RelativePath,
                Path = match.Path,
                Position = match.Position,
                Time = match.Time,
                Channel = match.Channel,
                Z = match.Z,
            });
        }

        var axes = new Dictionary<string, List<string>>
        {
            ["position"] = matches.Select(m => m.Position).Distinct().OrderBy(SortKey).ToList(),
            ["time"] = matches.Select(m => m.Time).Distinct().OrderBy(SortKey).ToList(),
            ["channel"] = matches.Select(m => m.Channel).Distinct().OrderBy(SortKey).ToList(),
            ["z"] = matches.Select(m => m.Z).Distinct().OrderBy(SortKey).ToList(),
        };
        var axisIndex = axes.ToDictionary(
            pair => pair.Key,
            pair => pair.Value.Select((value, index) => (value, index)).ToDictionary(v => v.value, v => v.index));

        return new SeriesDiscovery
        {
            Folder = folder,
            RecordCount = records.Count,
            Records = records,
            Axes = axes,
            Dataset = new SeriesDatasetPayload
            {
                Folder = folder,
                SubfolderTemplate = subfolderTemplate,
                FilenameTemplate = filenameTemplate,
                Axes = axes,
                AxisIndex = axisIndex,
                Lookup = lookup,
                Records = datasetRecords,
            },
        };
    }

    private static List<SeriesMatch> CollectMatches(string folder, string subfolderTemplate, string filenameTemplate)
    {
        var subfolderPattern = CompileTemplate(subfolderTemplate, allowEmpty: true);
        var filenamePattern = CompileTemplate(filenameTemplate);
        var matches = new List<SeriesMatch>();
        var candidateFolders = string.IsNullOrEmpty(subfolderTemplate)
            ? [new DirectoryInfo(folder)]
            : new DirectoryInfo(folder).GetDirectories().OrderBy(d => d.Name).ToArray();

        foreach (var candidate in candidateFolders)
        {
            var subfolderName = string.IsNullOrEmpty(subfolderTemplate) ? "" : candidate.Name;
            if (!string.IsNullOrEmpty(subfolderTemplate))
            {
                if (!subfolderPattern.IsMatch(subfolderName))
                    continue;
            }
            else if (!candidate.Exists)
            {
                continue;
            }

            var subfolderValues = subfolderPattern.Match(subfolderName).Groups
                .Cast<Group>()
                .Where(g => g.Name is "position" or "time" or "channel" or "z")
                .ToDictionary(g => g.Name, g => g.Value);

            foreach (var file in candidate.GetFiles().Where(f => IsSupportedSeriesFile(f.Name)).OrderBy(f => f.Name))
            {
                var filenameMatch = filenamePattern.Match(file.Name);
                if (!filenameMatch.Success)
                    continue;

                var filenameValues = filenameMatch.Groups
                    .Cast<Group>()
                    .Where(g => g.Name is "position" or "time" or "channel" or "z")
                    .ToDictionary(g => g.Name, g => g.Value);

                var values = new Dictionary<string, string>(subfolderValues);
                foreach (var pair in filenameValues)
                    values[pair.Key] = pair.Value;

                var relativePath = string.IsNullOrEmpty(subfolderName)
                    ? file.Name
                    : $"{subfolderName}/{file.Name}";

                matches.Add(new SeriesMatch
                {
                    Path = file.FullName,
                    RelativePath = relativePath,
                    Position = values.GetValueOrDefault("position", "0"),
                    Time = values.GetValueOrDefault("time", "0"),
                    Channel = values.GetValueOrDefault("channel", "0"),
                    Z = values.GetValueOrDefault("z", "0"),
                });
            }
        }

        return matches;
    }

    private static Regex CompileTemplate(string template, bool allowEmpty = false)
    {
        if (string.IsNullOrEmpty(template))
        {
            if (allowEmpty)
                return new Regex("^$", RegexOptions.IgnoreCase);
            throw new SidecarException("Filename template cannot be empty.");
        }

        var parts = new List<string> { "^" };
        var index = 0;
        var matches = Regex.Matches(template, @"\{([a-zA-Z_][a-zA-Z0-9_]*)\}");
        foreach (Match match in matches)
        {
            parts.Add(Regex.Escape(template[index..match.Index]));
            parts.Add(match.Groups[1].Value switch
            {
                "t" or "time" => "(?<time>.+?)",
                "p" or "position" => "(?<position>.+?)",
                "c" or "channel" => "(?<channel>.+?)",
                "z" => "(?<z>.+?)",
                _ => throw new SidecarException($"Unsupported placeholder {{{match.Groups[1].Value}}}"),
            });
            index = match.Index + match.Length;
        }

        parts.Add(Regex.Escape(template[index..]));
        parts.Add("$");
        return new Regex(string.Concat(parts), RegexOptions.IgnoreCase);
    }

    private static bool IsSupportedSeriesFile(string name)
    {
        if (name.EndsWith("_seg.cellpose", StringComparison.OrdinalIgnoreCase))
            return false;
        return SupportedExtensions.Contains(Path.GetExtension(name).ToLowerInvariant());
    }

    private static IComparable SortKey(string value) =>
        int.TryParse(value, out var number) ? number : value;

    private sealed class SeriesMatch
    {
        public string Path { get; init; } = "";
        public string RelativePath { get; init; } = "";
        public string Position { get; init; } = "0";
        public string Time { get; init; } = "0";
        public string Channel { get; init; } = "0";
        public string Z { get; init; } = "0";
    }
}
